### DISCUSSION TITLE
[Question] Question title Importing customized .usd file and make it to be rigidbody

### DISCUSSION THREAD

**Discussion by calmdw** [isaac-sim/IsaacLab#683] [Q&A]
### Question
HI, I have imported a .usd file and set it to be rigid body, but it can not be set as rigid body.

-> CODE:

cfg = sim_utils.UsdFileCfg(
        usd_path="~/Downloads/Cranfield.usd",
        scale=(0.001, 0.001, 0.001),
        rigid_props=sim_utils.RigidBodyPropertiesCfg(
            rigid_body_enabled=True,
            kinematic_enabled=True,
            disable_gravity=False,
            solver_position_iteration_count=8,
            solver_velocity_iteration_count=0,
            sleep_threshold=0.05,
            stabilization_threshold=0.0025,
            max_depenetration_velocity=1000.0,
        ),
    )

cfg.func("/World/Objects/objects", cfg, translation=(0.0, 0.0, 1.0))

->  WARING:
2024-07-10 13:52:02 [7,144ms] [Warning] [omni.isaac.lab.sim.utils] Could not perform 'modify_rigid_body_properties' on any prims under: '/World/Objects/objects'. This might be because of the following reasons:
	(1) The desired attribute does not exist on any of the prims.
	(2) The desired attribute exists on an instanced prim.
		Discovered list of instanced prim paths: ['/World/Objects/objects/Geometry/Cranfield_0/BolzenKlein_1_5', '/World/Objects/objects/Geometry/Cranfield_0/BolzenKlein_2_7', '/World/Objects/objects/Geometry/Cranfield_0/BolzenKleinEckig_1_8', '/World/Objects/objects/Geometry/Cranfield_0/BolzenKleinEckig_2_10']

**Comment by xxje-library** [comment]
Do you know how to solve the problem now

**Comment by mickeyouyou** [comment]
Do you have any latest progress?

**Comment by lopatovsky** [comment]
The failed function ['modify_rigid_body_properties'](https://github.com/isaac-sim/IsaacLab/blob/788a061d57ead17ff669eccce3d776fbccc59790/source/extensions/omni.isaac.lab/omni/isaac/lab/sim/schemas/schemas.py#L221) which is modifying the RigidObject some values specified in config  (e.g rigid_body_enabled, kinematic_enabled).
For this to work, you need to apply RigidBodyAPI to representation of your rigid object in the usd

**Comment by mmayank74567** [comment]
You must define Physics API in your USD (if not already present) to be considered a rigid object.

As mentioned in the `rigid_object.py` file within the IsaacLab directory, for an asset to be considered as a rigid object, the root prim of the asset must have the USD RigidBodyAPI applied to it.

The USD you are trying to import may not have physics API, so you can enable the following

- RigidBodyAPI
- MassAPI
- CollisionAPI

After adding these to the root prim, you can import the rigid body using the RigidObjectCfg using a USD file.

Adding below the code to add the Physics API to the USD root prim.

```
from pxr import Usd, UsdGeom, UsdPhysics, Gf

def analyze_usd_file(file_path):

    # Open the USD stage
    stage = Usd.Stage.Open(file_path)
    if not stage:
        print(f"Could not open USD file: {file_path}")
        return None


    print(f"Opened USD file: {file_path}")


    prim_data = {}
    for prim in stage.Traverse():

        prim_path = str(prim.GetPath())
        prim_type = prim.GetTypeName()
        prim_name = prim.GetName()
        print(f"Analyzing Prim: {prim_name} | Type: {prim_type} | Path: {prim_path}")


        properties = {} #use this section to see the properties of the USD
        for prop in prim.GetProperties():
            prop_name = prop.GetName()
            prop_value = prim.GetAttribute(prop_name).Get()
            print(f"Property: {prop_name} | Value: {prop_value}")

        rigid_body_api = UsdPhysics.RigidBodyAPI.Get(stage, prim_path)

        if rigid_body_api:
            print(f"The prim '{prim_path}' has RigidBodyAPI applied.")
        else:
            print(f"The prim '{prim_path}' does NOT have RigidBodyAPI applied.")
            if prim.CanApplyAPI(UsdPhysics.RigidBodyAPI):
                print("Applying RigidBodyAPI...")
                UsdPhysics.RigidBodyAPI.Apply(prim)
                print("RigidBodyAPI applied.")

        mass_api = UsdPhysics.MassAPI.Get(stage, prim_path)
        if mass_api:
            print(f"The prim '{prim_path}' has MassAPI applied.")
        else:
            print(f"The prim '{prim_path}' does NOT have MassAPI applied.")
            if prim.CanApplyAPI(UsdPhysics.MassAPI):
                print("Applying MassAPI...")
                UsdPhysics.MassAPI.Apply(prim)
                print("MassAPI applied.")


        collision_api = UsdPhysics.CollisionAPI.Get(stage, prim_path)
        if collision_api:
            print(f"The prim '{prim_path}' has CollisionAPI applied.")
        else:
            print(f"The prim '{prim_path}' does NOT have CollisionAPI applied.")
            if prim.CanApplyAPI(UsdPhysics.CollisionAPI):
                print("Applying CollisionAPI...")
                UsdPhysics.CollisionAPI.Apply(prim)
                print("CollisionAPI applied.")
        stage.GetRootLayer().Save()
        break # as we are just doing it for the root


    return prim_data


def main():

    usd_file_path = "<your_path>.usd"

    prim_data = analyze_usd_file(usd_file_path)



if __name__ == "__main__":
    main()
```
